src/algebra/lie/from_cartan_matrix.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.lie.free
+import algebra.lie.quotient
+
+/-!
+# Lie algebras from Cartan matrices
+
+Split semi-simple Lie algebras are uniquely determined by their Cartan matrix. Indeed, if `A` is
+an `l × l` Cartan matrix, the corresponding Lie algebra may be obtained as the Lie algebra on
+`3l` generators: $H_1, H_2, \ldots H_l, E_1, E_2, \ldots, E_l, F_1, F_2, \ldots, F_l$
+subject to the following relations:
+$$
+\begin{align}
+  [H_i, H_j] &= 0\\
+  [E_i, F_i] &= H_i \quad\mbox{ and }\quad [E_i, F_j] = 0 \quad\mbox{if $i ≠ j$}\\
+  [H_i, E_j] &= A_{ij}E_j\\
+  [H_i, F_j] &= -A_{ij}F_j\\
+  ad(E_i)^{1 - A_{ij}}(E_j) &= 0 \quad\mbox{if $i \ne j$}\\
+  ad(F_i)^{1 - A_{ij}}(F_j) &= 0 \quad\mbox{if $i \ne j$}\\
+\end{align}
+$$
+
+In this file we provide the above construction. It is defined for any matrix of integers but the
+results for non-Cartan matrices should be regarded as junk.
+
+## Main definitions
+
+  * `lie_algebra.from_cartan_matrix`
+
+## Tags
+
+lie algebra, semi-simple, cartan matrix
+-/
+
+universes u v w
+
+noncomputable theory
+
+variables (R : Type u) {B : Type v} [comm_ring R] [decidable_eq B] [fintype B]
+variables (cartan_matrix : matrix B B ℤ)
+
+namespace lie_algebra
+
+namespace from_cartan_matrix
+
+variables (B)
+
+/-- The generators of the free Lie algebra from which we construct the Lie algebra of a Cartan
+matrix as a quotient. -/
+inductive generators
+| H : B → generators
+| E : B → generators
+| F : B → generators
+
+instance [inhabited B] : inhabited (generators B) := ⟨generators.H $ default B⟩
+
+variables {B}
+
+namespace relations
+
+local notation `H` := free_lie_algebra.of R ∘ generators.H
+local notation `E` := free_lie_algebra.of R ∘ generators.E
+local notation `F` := free_lie_algebra.of R ∘ generators.F
+local notation `ad` := lie_algebra.ad R (free_lie_algebra R (generators B))
+
+/-- The `⁅H, H⁆`-relation. -/
+def HH : B × B → free_lie_algebra R (generators B) := function.uncurry $ λ i j,
+⁅H i, H j⁆
+
+/-- The `⁅E, F⁆`-relation. -/
+def EF : B × B → free_lie_algebra R (generators B) := function.uncurry $ λ i j,
+if i = j then ⁅E i, F i⁆ - H i else ⁅E i, F j⁆
+
+/-- The `⁅H, E⁆`-relation. -/
+def HE : B × B → free_lie_algebra R (generators B) := function.uncurry $ λ i j,
+⁅H i, E j⁆ - (cartan_matrix i j) • E j
+
+/-- The `⁅H, F⁆`-relation. -/
+def HF : B × B → free_lie_algebra R (generators B) := function.uncurry $ λ i j,
+⁅H i, F j⁆ + (cartan_matrix i j) • F j
+
+/-- The `ad E`-relation.
+
+Note that we use `int.to_nat` so that we can take the power and that we do not bother
+restricting to the case `i ≠ j` since these relations are zero anyway. We also defensively
+ensure this with `ad_E_of_eq_eq_zero`. -/
+def ad_E : B × B → free_lie_algebra R (generators B) := function.uncurry $ λ i j,
+(ad (E i))^(-cartan_matrix i j).to_nat $ ⁅E i, E j⁆
```

Comment:
```suggestion
def ad_E : B × B → free_lie_algebra R (generators B) :=
function.uncurry $ λ i j, (ad (E i))^(-cartan_matrix i j).to_nat $ ⁅E i, E j⁆
```

Reply:
Or maybe you thought the initial code is stylistically better? Because it puts the focus on the actual content?

Reply:
I was indeed attempting to focus on the actual content but I recognise I'm diverging from offical style. I'm 50-50 on which is better. What do you prefer?

Reply:
Maybe following the official style is less awkard after `open function`?